storage/tokudb/hatoku_cmp.cc

```diff
@@ -1623,11 +1623,298 @@ static uint32_t create_toku_main_key_pack_descriptor(uchar *buf) {
+  }
+  {
+    uint32_t curr_fixed_offset = 0;
+    uint32_t curr_var_index = 0;
+    for (uint32_t j = 0; j < table_share->fields; j++) {
+      COL_PACK_INFO *curr = &cp_info[keynr][j];
+      //
+      // need to set the offsets / indexes
+      // offsets are calculated AFTER the NULL bytes
+      //
+      if (!bitmap_is_set(&key_filters[keynr], j)) {
+        if (is_fixed_field(j)) {
+          curr->col_pack_val = curr_fixed_offset;
+          curr_fixed_offset += field_lengths[j];
+        } else if (is_variable_field(j)) {
+          curr->col_pack_val = curr_var_index;
+          curr_var_index++;
+        }
+      }
+    }
+
+    //
+    // set up the mcp_info
+    //
+    mcp_info[keynr].fixed_field_size = get_fixed_field_size(table_share, keynr);
+    mcp_info[keynr].len_of_offsets = get_len_of_offsets(table_share, keynr);
+
+    error = 0;
+  }
+exit:
+  return error;
+}
+
+void KEY_AND_COL_INFO::free() {
+  if (allocated) {
+    for (uint32_t i = 0; i < MAX_KEY + 1; i++) {
+      bitmap_free(&key_filters[i]);
+    }
+
+    for (uint32_t i = 0; i < MAX_KEY + 1; i++) {
+      tokudb::memory::free(cp_info[i]);
+      cp_info[i] = NULL;  // 3144
+    }
+
+    tokudb::memory::free(multi_ptr);
+    field_types = NULL;
+    field_lengths = NULL;
+    length_bytes = NULL;
+    blob_fields = NULL;
+    num_blobs = 0;
+    memset(&mcp_info, 0, sizeof(mcp_info));
+    memset(&cp_info, 0, sizeof(cp_info));
+    num_offset_bytes = 0;
+
+    allocated = false;
+  }
+}
+
+void KEY_AND_COL_INFO::reset(uint32_t keynr) {
+  bitmap_clear_all(&key_filters[keynr]);
+  tokudb::memory::free(cp_info[keynr]);
+  cp_info[keynr] = NULL;
+  mcp_info[keynr] = (MULTI_COL_PACK_INFO){0, 0};
+}
+
+uint32_t KEY_AND_COL_INFO::get_fixed_field_size(TABLE_SHARE *table_share,
+                                                uint32_t keynr) const {
+  uint32_t offset = 0;
+  for (uint32_t i = 0; i < table_share->fields; i++) {
+    if (is_fixed_field(i) && !bitmap_is_set(&key_filters[keynr], i)) {
+      offset += field_lengths[i];
+    }
+  }
+  return offset;
+}
+
+uint32_t KEY_AND_COL_INFO::get_len_of_offsets(TABLE_SHARE *table_share,
+                                              uint32_t keynr) const {
+  uint32_t len = 0;
+  for (uint32_t i = 0; i < table_share->fields; i++) {
+    if (is_variable_field(i) && !bitmap_is_set(&key_filters[keynr], i)) {
+      len += num_offset_bytes;
+    }
+  }
+  return len;
+}
+
+uint32_t KEY_AND_COL_INFO::get_max_desc_size(TABLE *form) const {
+  uint32_t max_row_desc_buff_size;
+  // upper bound of key comparison descriptor
+  max_row_desc_buff_size = 2 * (form->s->fields * 6) + 10;
+  // upper bound for sec. key part
+  max_row_desc_buff_size += get_max_secondary_key_pack_desc_size();
+  // upper bound for clustering val part
+  max_row_desc_buff_size += get_max_clustering_val_pack_desc_size(form->s);
+  return max_row_desc_buff_size;
+}
+
+uint32_t KEY_AND_COL_INFO::get_max_secondary_key_pack_desc_size() const {
```

Comment:
Mark inline, and move to header, it seems it should compile to "return a constant plus a bit". Might as well mark as noexcept.